specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+
+An "Index offset" value of `0` in the CARv2 header indicates that there is no index in this CARv2 container and no attempt should be made to read it.
+
+### Index format
+
+The CARv2 index is a flexible container format itself, allowing for different index layouts depending on suitability for a particular application or set of data - generation speed, usage performance, size, etc. Supported index formats will be detailed in this specification, below.
+
+Index data, once read from the CARv2 container, provides a mapping of hash digest bytes to block location in byte offset from the beginning of the CARv2 container. The index only uses the hash digest&mdash;it does not use the full bytes of a CID, nor does it use any of the multihash prefix bytes.
+
+The first byte(s) of a CARv2 index are a varint that indicates the index format type. The remaining bytes follow the encoding rules of that index format type.
+
+As the index only contains the hash digest bytes, other details contained within the block's CID and the length of the block's bytes must be derived by inspecting the initial bytes of the block entry within the data payload.
+
+**TODO:** How are we handling identity hash CIDs?
+* Do we want to state that identity multihash CIDs are explicitly excluded from indexes? 
+* Can we defer an assumption that a block store interface would return the bytes without having to look them up when it sees an identity CID? 
+* What are the implications of having an index that doesn't contain all of the blocks within the CAR?
+* Or do we want to explicitly exclude index CIDs entirely from the CARv2 format since they are arguably a waste of space?
+
+#### Format `0x300000`: IndexSorted
+
+**TODO:** I'm assuming this is supposed to be a multicodec and that `0x300000` is using the reserved range - so we should register this in the table.
+
+An index format type of `0x300000` indicates the "IndexSorted" format.
+
+IndexSorted sorts hash digests by two dimensions: first into buckets of _digest length_, smallest to largest, and then within those buckets ordered by a simple byte-wise sorting. In this way, locating a hash digest within the CAR requires first finding the bucket matching the length of the requested hash digest, then searching the ordered list of digests within that bucket to find the matching entry.
+
+* IndexSorted may contain one or more length-grouped buckets of digests.
+* Buckets are ordered by digest length and concatenated together to form the index.
+* Each bucket is prefixed with:
+  * a "width" encoded as a 32-bit **signed** little-endian integer indicating the common byte length of the combination of hash digests and their offsets in this bucket; followed by
+  * a "count" encoded as a 64-bit **signed** little-endian integer which determines the total number of hash digests (and their offsets) bucket.
```

Comment:
How come these two are signed? They could be unsigned from what I can tell, and then one doesn't have to worry about negative values.

Reply:
cause they are signed currently, I bolded them to highlight this fact and assumed we'd get agreement to change that: https://github.com/ipld/go-car/blob/2a49cbd958b0c6443dd1e9b518814199bf50842d/v2/internal/index/indexsorted.go#L21-L22

@masih maybe let me know when they're changed and I'll update here, I just don't want to risk the spec & impl being out of sync.

Reply:
Thanks for pointing this out @rvagg

Reply:
This pr should fix it https://github.com/ipld/go-car/pull/96

Reply:
doc updated